pkg/controllers/proxyconfig/proxyconfig_controller.go

```diff
@@ -0,0 +1,134 @@
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	"golang.org/x/net/http/httpproxy"
+
+	"k8s.io/klog/v2"
+)
+
+// proxyConfigChecker reports bad proxy configurations.
+type proxyConfigChecker struct {
+	proxyConfig *httpproxy.Config
+	routeLister v1.RouteLister
+}
+
+func NewProxyConfigChecker(
+	routeInformer routeinformer.RouteInformer,
+	recorder events.Recorder,
+	operatorClient v1helpers.OperatorClient) factory.Controller {
+	p := proxyConfigChecker{
+		proxyConfig: httpproxy.FromEnvironment(),
+		routeLister: routeInformer.Lister(),
+	}
+
+	return factory.New().
+		WithSync(p.sync).
+		WithInformers(routeInformer.Informer()).
+		ResyncEvery(30*time.Second).
+		WithSyncDegradedOnError(operatorClient).
+		ToController("ProxyConfigController", recorder.WithComponentSuffix("proxy-config-controller"))
+}
+
+func (p *proxyConfigChecker) sync(ctx context.Context, _ factory.SyncContext) error {
+	if !p.isProxyConfigured() {
+		// If proxy is not configured, then it is a no-op.
+		return nil
+	}
+
+	route, err := p.routeLister.Routes("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		return err
+	}
+
+	endpointURL, err := url.Parse(route.Spec.Host)
+	if err != nil {
+		return err
+	}
+
+	withProxySuccess := p.isEndpointReachable(ctx, "https://"+route.Spec.Host+"/healthz", p.proxyFunc)
+	withoutProxySuccess := p.isEndpointReachable(ctx, "https://"+route.Spec.Host+"/healthz", nil)
+	noProxyMatchesEndpoint := strings.Contains(p.proxyConfig.NoProxy, endpointURL.Hostname())
+
+	if noProxyMatchesEndpoint && !withoutProxySuccess {
+		if withProxySuccess {
+			return fmt.Errorf("endpoint(%q) found in NO_PROXY(%q) but can only be reached with proxy", endpointURL.Hostname(), p.proxyConfig.NoProxy)
+		}
+		return fmt.Errorf("endpoint(%q) found in NO_PROXY(%q) is not reachable with/without proxy", endpointURL.Hostname(), p.proxyConfig.NoProxy)
+	}
+
+	if !noProxyMatchesEndpoint && !withProxySuccess {
+		if withoutProxySuccess {
+			return fmt.Errorf("endpoint(%q) absent in NO_PROXY(%q) can only be reached without proxy", endpointURL.Hostname(), p.proxyConfig.NoProxy)
+		}
+		return fmt.Errorf("endpoint(%q) is not reachable with/without proxy", endpointURL.Hostname())
+	}
+
+	return nil
+}
+
+func (p *proxyConfigChecker) isEndpointReachable(ctx context.Context, endpointURL string, proxyFunc func(req *http.Request) (*url.URL, error)) bool {
+	req, err := http.NewRequest(http.MethodGet, endpointURL, nil)
+	if err != nil {
+		klog.V(4).Infof("failed to create new http request: %v", err)
+		return false
+	}
+	reqCtx, cancel := context.WithTimeout(ctx, 10*time.Second) // avoid waiting forever
+	defer cancel()
+	req.WithContext(reqCtx)
+
+	client := &http.Client{
+		Timeout: 5 * time.Second,
+		Transport: &http.Transport{
+			Proxy: proxyFunc,
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
```

Comment:
what if the CA in proxyconfig is misconfigured?

Reply:
Good point. Now it uses ca to validate proxyconfig.